// rehype-mdx-title/index.js
import { name as isIdentifierName } from 'estree-util-is-identifier-name';
import { toString } from 'hast-util-to-string';
import { EXIT, visit } from 'unist-util-visit';
/**
 * A rehype plugin to expose the MDX page title as string.
 *
 * @param options Optional options to configure the output.
 * @returns A unified transformer.
 */
const rehypeMdxTitle = ({ name = 'title' } = {}) => {
    if (!isIdentifierName(name)) {
        throw new Error(`The name should be a valid identifier name, got: ${JSON.stringify(name)}`);
    }
    return (ast) => {
        visit(ast, { type: 'element', tagName: 'h1' }, (node) => {
            const value = toString(node);
            ast.children.unshift({
                type: 'mdxjsEsm',
                value: '',
                data: {
                    estree: {
                        type: 'Program',
                        sourceType: 'module',
                        body: [
                            {
                                type: 'ExportNamedDeclaration',
                                source: null,
                                specifiers: [],
                                declaration: {
                                    type: 'VariableDeclaration',
                                    kind: 'const',
                                    declarations: [
                                        {
                                            type: 'VariableDeclarator',
                                            id: { type: 'Identifier', name },
                                            init: {
                                                type: 'Literal',
                                                value,
                                                raw: JSON.stringify(value),
                                            },
                                        },
                                    ],
                                },
                            },
                        ],
                    },
                },
            });
            return EXIT;
        });
    };
};
export default rehypeMdxTitle;
